Issue/3785 - Add alignment controls to heading block toolbar

amdrew opened the pull request:
## Description

Adds the alignment controls to the toolbar, as discussed in #3785

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.

amdrew committed:
Add alignment controls to toolbar #3785

aduth commented on blocks/library/heading/index.js:
```diff
@@ -117,7 +126,9 @@ export const settings = {
 					}
-				/>
+				>
+				{ alignmentToolbar }
```
Can we tab this once more? (Ideally we should have an ESLint rule for this)

amdrew replied:
Fixed

aduth reviewed (commented):
Code-wise looks good.

>It seems we should also remove the alignment controls from the block inspector? If so, I'll update the PR.

I'd like some feedback here. Seems to me at least we'd want to drop the duplicated controls from the sidebar inspector. cc @jasmussen

amdrew committed:
Indent { alignmentToolbar }

jasmussen commented:
Thanks so much for working on this! Here's a screenshot:

<img width="773" alt="screen shot 2018-04-03 at 08 36 00" src="https://user-images.githubusercontent.com/1204802/38233261-46b6a324-371a-11e8-8125-1830f7d79a34.png">

This looks good, but it doesn't account for responsive or mobile breakpoints:

<img width="526" alt="screen shot 2018-04-03 at 08 36 06" src="https://user-images.githubusercontent.com/1204802/38233273-4f223d52-371a-11e8-8ed1-aa69be1fb892.png">

<img width="434" alt="screen shot 2018-04-03 at 08 37 13" src="https://user-images.githubusercontent.com/1204802/38233276-50ceac4e-371a-11e8-8eb7-0d65d5d2ab66.png">

One suggestion detailed in https://github.com/WordPress/gutenberg/issues/3785#issuecomment-351412573 is to remove the 3 quick shortcuts to set h2 h3 and h4, and instead move those to the switcher interface. To here:

<img width="417" alt="screen shot 2018-04-03 at 08 37 28" src="https://user-images.githubusercontent.com/1204802/38233317-71c7cea8-371a-11e8-80ea-578e651e059f.png">

☝️ Imagine headings 1-6 all in that list. 

I realize that's a lot more work than just adding the alignments toolbar, and if this proves too much work for you then that's totally fine, we'll see if one of the core devs has time to work on it. 

I also agree with Andrew, we should remove both the heading level and alignments from the inspector:

<img width="243" alt="screen shot 2018-04-03 at 08 39 34" src="https://user-images.githubusercontent.com/1204802/38233399-bbbce412-371a-11e8-9e6d-87fd67a56faa.png">

amdrew committed:
Merge branch 'master' of github.com:WordPress/gutenberg into issue/3785

amdrew committed:
Remove alignment toolbar from block controls

amdrew committed:
Remove const, move AlignmentToolbar

amdrew committed:
Fix eslint warnings about indentation

amdrew commented:
@jasmussen I've updated the PR and removed the alignment controls from the block inspector. 

Good spotting with the responsiveness, I hadn't considered that. Is placing other inspector controls inside the transform menu a pattern that Gutenberg utilizes with other blocks at mobile device widths? 

Though it seems you can scroll horizontally to reach other inspector controls (only tested on iOS simulator for mac), I'm unaware on how other mobile devices treat this - Do they all provide some mechanism to scroll horizontally to reach the other controls? If not, moving the headings into the transform menu does seem like the best compromise.

In any case, I've left the headings alone for now as I'm not sure how to move them inside the transform menu, and only for mobile devices (is there a pre-existing function for this?). If someone can give me some pointers I'll definitely give it a shot.

jasmussen commented:
> Good spotting with the responsiveness, I hadn't considered that. Is placing other inspector controls inside the transform menu a pattern that Gutenberg utilizes with other blocks at mobile device widths?

In general nothing should be put inside the inspector that is necessary for the basic operation of the block, either on desktop or mobile. In this case, the PR addresses the issue that _alignments_ could be considered necessary for the basic operation of the heading block. I don't disagree with that, and there is consensus that this is the case. 

The horizontal scrollbar on mobile is a "last resort" at making sure all controls are available, but it's not a good pattern and it's not scalable. The responsive desktop breakpoint is more of an issue, because you can't scroll there.

There are ideas of putting alignments inside a dropdown, but that's also a bandaid.

The best solution is to be careful what we put in the block toolbar, and make sure it scales well. In this case, I thin the solution is, as previously noted, to remove the headings from the block toolbar and add them as transformations.

Thank you for your work so far, really appreciate it. @gziolo if you have bandwidth, do you think you can look at putting the heading sizes inside the switcher? And @karmatosed do you agree that's the best approach for now?

amdrew commented:
Got it, I agree. It's difficult to scroll horizontally on a mobile device unless you know you actually can in the first place.

jasmussen commented:
So, here are mockups because they speak louder than words. It's not working quite as well visually, as I'd hoped, but it can serve as a basis for discussion.

![heading block toolbar](https://user-images.githubusercontent.com/1204802/38304242-fe73a7de-3808-11e8-88fb-8d8f57111b47.png)

![transformations](https://user-images.githubusercontent.com/1204802/38304247-01427e36-3809-11e8-9160-256dadbb4aa4.png)

I imagine you'd still want to have just a single heading block that you inserted, which would default to H2.

gziolo commented:
The same applies to other blocks:

![screen shot 2018-04-04 at 13 32 57](https://user-images.githubusercontent.com/699132/38305314-bdee1fa6-380c-11e8-9af3-e317e2c6402b.png)

![screen shot 2018-04-04 at 13 33 28](https://user-images.githubusercontent.com/699132/38305329-c72afb84-380c-11e8-8c94-bccd4db93763.png)

karmatosed commented:
@jasmussen I like these mocks. I do wonder if that's a lot of text in a line and the 'size' is needed, maybe it is. I do like it better than what we have now and think it's worth creating to experience.
